internal/provisioner/helm/controllers/bundledeployment_controller.go

```diff
@@ -0,0 +1,480 @@
+				Message: err.Error(),
+			}))
+		return ctrl.Result{}, err
+	}
+
+	rel, state, err := r.getReleaseState(cl, bd, chrt)
+	if err != nil {
+		u.UpdateStatus(
+			updater.EnsureCondition(metav1.Condition{
+				Type:    rukpakv1alpha1.TypeInstalled,
+				Status:  metav1.ConditionFalse,
+				Reason:  rukpakv1alpha1.ReasonErrorGettingReleaseState,
+				Message: err.Error(),
+			}))
+		return ctrl.Result{}, err
+	}
+
+	switch state {
+	case stateNeedsInstall:
+		_, err = cl.Install(bd.Name, r.ReleaseNamespace, chrt, values, func(install *action.Install) error {
+			install.CreateNamespace = false
+			return nil
+		})
+		if err != nil {
+			if isResourceNotFoundErr(err) {
+				err = errRequiredResourceNotFound{err}
+			}
+			u.UpdateStatus(
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeInstalled,
+					Status:  metav1.ConditionFalse,
+					Reason:  rukpakv1alpha1.ReasonInstallFailed,
+					Message: err.Error(),
+				}))
+			return ctrl.Result{}, err
+		}
+	case stateNeedsUpgrade:
+		_, err = cl.Upgrade(bd.Name, r.ReleaseNamespace, chrt, values)
+		if err != nil {
+			if isResourceNotFoundErr(err) {
+				err = errRequiredResourceNotFound{err}
+			}
+			u.UpdateStatus(
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeInstalled,
+					Status:  metav1.ConditionFalse,
+					Reason:  rukpakv1alpha1.ReasonUpgradeFailed,
+					Message: err.Error(),
+				}))
+			return ctrl.Result{}, err
+		}
+	case stateUnchanged:
+		if err := cl.Reconcile(rel); err != nil {
+			if isResourceNotFoundErr(err) {
+				err = errRequiredResourceNotFound{err}
+			}
+			u.UpdateStatus(
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeInstalled,
+					Status:  metav1.ConditionFalse,
+					Reason:  rukpakv1alpha1.ReasonReconcileFailed,
+					Message: err.Error(),
+				}))
+			return ctrl.Result{}, err
+		}
+	default:
+		return ctrl.Result{}, fmt.Errorf("unexpected release state %q", state)
+	}
+
+	u.UpdateStatus(
+		updater.EnsureCondition(metav1.Condition{
+			Type:    rukpakv1alpha1.TypeInstalled,
+			Status:  metav1.ConditionTrue,
+			Reason:  rukpakv1alpha1.ReasonInstallationSucceeded,
+			Message: fmt.Sprintf("instantiated bundle %s successfully", bundle.GetName()),
+		}),
+		updater.EnsureInstalledName(bundle.GetName()),
+	)
+	if err := r.reconcileOldBundles(ctx, bundle, allBundles); err != nil {
+		l.Error(err, "failed to delete old bundles")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcileDesiredBundle is responsible for checking whether the desired
+// Bundle resource that's specified in the BundleDeployment parameter's
+// spec.Template configuration is present on cluster, and if not, creates
+// a new Bundle resource matching that desired specification.
+func reconcileDesiredBundle(ctx context.Context, c client.Client, bd *rukpakv1alpha1.BundleDeployment) (*rukpakv1alpha1.Bundle, *rukpakv1alpha1.BundleList, error) {
```

Comment:
This function seems ripe for moving to a package that can be shared by plain and helm provisioners.

Reply:
I'll move this to internal/util/util.go